### PR TITLE
feat: add nok ferry data

### PR DESCRIPTION
**Pull request by lukashass**
## TODO

- [x] where to store/get the gtfs data?
- [ ] ~~only publish departures on subscription~~ later

### CODE REVIEW

**Comment by anbraten** [comment]
vehicle-ferry
vehicle-ferry-selected

File: go.mod [hunk 1]
```diff
@@ -13,6 +13,8 @@ require (
 
 require (
 	github.com/adrianmo/go-nmea v1.8.0 // indirect
+	github.com/artonge/go-csv-tag/v2 v2.2.0 // indirect
+	github.com/artonge/go-gtfs v1.3.4 // indirect
```
Comment (anbraten) [inline go.mod:17]: shouldn't be an indirect one. might be fixed by `go mod tidy`

File: collectors/kvg/collector/stop.go [hunk 2]
```diff
@@ -96,8 +96,8 @@ func (c *StopCollector) publishRemoved(stop *protocol.Stop) error {
 func (c *StopCollector) SubjectsToIDs(subjects []string) []string {
 	var ids []string
 	for _, subject := range subjects {
-		if strings.HasPrefix(subject, fmt.Sprintf(protocol.SubjectMapStop, "")) && subject != fmt.Sprintf(protocol.SubjectMapStop, ">") {
-			ids = append(ids, strings.TrimPrefix(subject, fmt.Sprintf(protocol.SubjectMapStop, "")+api.IDPrefix))
+		if strings.HasPrefix(subject, fmt.Sprintf(protocol.SubjectMapStop, api.IDPrefix)) && subject != fmt.Sprintf(protocol.SubjectMapStop, ">") {
```
Comment (anbraten) [inline collectors/kvg/collector/stop.go:99]: Do we have to adjust this for routes, etc as well?

File: app/src/components/popups/FerryStopPopup.vue [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+  if (arrival.state === 'stopping') {
+    return t('stopping');
+  }
+  if (minutes < 1) {
+    return t('immediately');
+  }
+
+  return t('minutes', { minutes });
+};
```
Comment (anbraten) [inline app/src/components/popups/FerryStopPopup.vue:115]: Is this sth from the bus-stop that we could move to central place?

File: collectors/gtfs/gtfs.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func loadRemoteGTFS(url string) (*gtfs.GTFS, error) {
+	// download file
+	err := downloadFile(url, "/tmp/gtfs-collector.zip")
```
Comment (anbraten) [inline collectors/gtfs/gtfs.go:83]: Use `os.CreateTemp()` instead

File: collectors/gtfs/main.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+		return
+	}
+	defer func() {
+		err := c.Disconnect()
+		if err != nil {
+			log.Error(err)
+		}
+	}()
+
+	g, err := loadRemoteGTFS("https://github.com/lukashass/nok-gtfs/raw/main/adler.zip")
```
Comment (anbraten) [inline collectors/gtfs/main.go:54]: Would be nice to have this as an env var to use the collector for other gtfs feeds later on. As it seems we would have to adjust some other parts as well, maybe also sth for a followup PR.

File: collectors/gtfs/main.go [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+		if !c.IsConnected() {
+			return
+		}
+
+		for _, gtfsStop := range g.Stops {
+			// convert to protocol.Stop
+			stop := protocol.Stop{
+				ID:   IDPrefix + gtfsStop.ID,
+				Name: gtfsStop.Name,
+				Type: protocol.StopTypeFerryStop,
```
Comment (anbraten) [inline collectors/gtfs/main.go:72]: Should be possible to detect using: https://github.com/artonge/go-gtfs/blob/master/const.go#L9

File: collectors/gtfs/main.go [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+				Location: protocol.Location{
+					Latitude:  int(gtfsStop.Latitude * 3600000),
+					Longitude: int(gtfsStop.Longitude * 3600000),
+				},
+				Alerts: []string{"Die Abfahrtszeiten können sich je nach Witterung oder Verkehrslage auf dem Nord-Ostsee-Kanal geringfügig verschieben. Die Verschiebung einer Fahrt dient der Sicherheit des Fahrbetriebes. Bei Ausfall der Fähre ist ein Busersatzverkehr eingerichtet."},
```
Comment (anbraten) [inline collectors/gtfs/main.go:77]:
```suggestion
				// TODO: get alerts from gtfs feed
				Alerts: []string{"Die Abfahrtszeiten können sich je nach Witterung oder Verkehrslage auf dem Nord-Ostsee-Kanal geringfügig verschieben. Die Verschiebung einer Fahrt dient der Sicherheit des Fahrbetriebes. Bei Ausfall der Fähre ist ein Busersatzverkehr eingerichtet."},
```

File: collectors/gtfs/main.go [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+			}
+			// for each trip remove stop times with the highest stop_sequence (last stop)
+			filteredStopTimes := make([]gtfs.StopTime, 0)
+			for _, stopTime := range g.StopsTimes {
```
Comment (anbraten) [inline collectors/gtfs/main.go:81]: If I get this correctly this is `O(n*n)`? Couldn't we create a filtered index from the stop time first and assign that index to the corresponding stop after that?

File: collectors/kvg/collector/stop.go (hunk 2 above)
Comment (lukashass) [reply collectors/kvg/collector/stop.go:99]: Only stops and trips, vehicles have a dummy implementation for this function.

File: collectors/gtfs/main.go (hunk 6 above)
Comment (lukashass) [reply collectors/gtfs/main.go:72]: It is possible, but not necessary now.

**Review by anbraten — APPROVED** [review]

**Pull request merged**